File: juros.zig
// Cálculo do juros, sendo que precisa de arrays pra isso
// Versão 0.1: 17/06/2024: versão feita sem muito conhecimento de Zig

const std = @import("std");
const math = std.math;

// estutura básica para simplificar as chamadas
const Juros = struct {
    quantidade: i16,
    composto: bool,
    periodo: f64,
    pagamentos: std.ArrayList(f64), // alterar esses valores
    pesos: std.ArrayList(f64), // para sua necessidade

    // calcula a somatória de pesos[]
    pub fn getPesoTotal(self: Juros) f64 {
        var acumulador: f64 = 0.0;
        var indice: usize = 0;
        while (indice < self.quantidade) {
            acumulador += self.pesos.items[indice];
            indice += 1;
        }
        return acumulador;
    }

    // calcula o acréscimo a partir dos juros e parcelas
    pub fn jurosParaAcrescimo(self: Juros, juros: f64) f64 {
        const pesoTotal = self.getPesoTotal();
        if (juros <= 0.0 or self.quantidade < 1 or self.periodo <= 0.0 or pesoTotal <= 0.0) {
            return 0.0;
        }
        var acumulador: f64 = 0.0;
        var indice: usize = 0;

        while (indice < self.quantidade) {
            if (self.composto) {
                acumulador += self.pesos.items[indice] / math.pow(f64, 1 + juros / 100.0, self.pagamentos.items[indice] / self.periodo);
            } else {
                acumulador += self.pesos.items[indice] / (1 + juros / 100.0 * self.pagamentos.items[indice] / self.periodo);
            }
            indice += 1;
        }

        return (pesoTotal / acumulador - 1.0) * 100.0;
    }

    // calcula os juros a partir do acréscimo e parcelas
    pub fn acrescimoParaJuros(self: Juros, acrescimo: f64, precisao: f64, maxIteracoes: i16, maximoJuros: f64) f64 {
        const pesoTotal = self.getPesoTotal();
        if (acrescimo <= 0.0 or self.quantidade < 1 or self.periodo <= 0.0 or pesoTotal <= 0.0 or precisao < 1 or maxIteracoes < 1 or maximoJuros <= 0.0) {
            return 0.0;
        }
        var minJuros: f64 = 0.0;
        var medJuros: f64 = maximoJuros / 2.0;
        var maxJuros: f64 = maximoJuros;
        const minDiferenca: f64 = math.pow(f64, 0.1, precisao);
        var indice: usize = 0;

        while (indice < maxIteracoes) {
            medJuros = (minJuros + maxJuros) / 2.0;
            if ((maxJuros - minJuros) < minDiferenca) {
                return medJuros;
            }
            if (self.jurosParaAcrescimo(medJuros) < acrescimo) {
                minJuros = medJuros;
            } else {
                maxJuros = medJuros;
            }

            indice += 1;
        }

        return medJuros;
    }
};

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    const alocador = std.heap.page_allocator;
    const quantidade = 3;

    // cria os arrays de pagamentos e pesos
    var pagamentos = std.ArrayList(f64).init(alocador);
    var pesos = std.ArrayList(f64).init(alocador);
    var indice: f64 = 0;
    while (indice < quantidade) {
        try pagamentos.append((indice + 1.0) * 30.0);
        try pesos.append(1.0);
        indice += 1;
    }

    // cria um objeto oJuros do tipo Juros e incializa os valores
    const oJuros = Juros{ .quantidade = quantidade, .composto = true, .periodo = 30.0, .pagamentos = pagamentos, .pesos = pesos };

    // calcula e guarda os retornos das funções
    const pesoTotal: f64 = oJuros.getPesoTotal();
    const acrescimoCalculado: f64 = oJuros.jurosParaAcrescimo(3.0);
    const jurosCalculado: f64 = oJuros.acrescimoParaJuros(acrescimoCalculado, 15.0, 100, 50);

    // imprime os resultados
    try stdout.print("Peso total = {d}\n", .{pesoTotal});
    try stdout.print("Acrescimo = {d}\n", .{acrescimoCalculado});
    try stdout.print("Juros = {d}\n", .{jurosCalculado});
}

File: juros.ex
-- Cálculo dos juros, sendo que precisa de parcelas pra isso
-- Versão 0.1: 05/07/2024: versão feita sem muito conhecimento de Euphoria

include std/io.e

-- propriedades globais para simplificar as chamadas
global atom Quantidade = 3
global atom Composto = 1
global atom Periodo = 30.0
global sequence Pagamentos = {30.0, 60.0, 90.0}
global sequence Pesos = {1.0, 1.0, 1.0}

-- calcula a somatória da sequência Pesos[]
function getPesoTotal()
    atom acumulador = 0.0
    for indice = 1 to Quantidade do
        acumulador += Pesos[indice]
    end for
    return acumulador
end function

-- calcula o acréscimo a partir dos juros e parcelas
function jurosParaAcrescimo(atom juros)
    atom pesoTotal = getPesoTotal()
    if juros <= 0.0 or Quantidade < 1 or Periodo <= 0.0 or pesoTotal <= 0.0 then
        return 0.0
    end if
    atom acumulador = 0.0

    for indice = 1 to Quantidade do
        if Composto > 0 then
            acumulador += Pesos[indice] / power(1.0 + juros / 100.0, Pagamentos[indice] / Periodo)
        else
            acumulador += Pesos[indice] / (1.0 + juros / 100.0 * Pagamentos[indice] / Periodo)
        end if
    end for

    return (pesoTotal / acumulador - 1.0) * 100.0
end function

-- calcula os juros a partir do acréscimo e parcelas
function acrescimoParaJuros(atom acrescimo, atom precisao, atom maxIteracoes, atom maxJuros)
    atom pesoTotal = getPesoTotal()
    if acrescimo <= 0.0 or Quantidade < 1 or Periodo <= 0.0 or pesoTotal <= 0.0 or precisao < 1 or maxIteracoes < 1 or maxJuros <= 0.0 then
        return 0.0
    end if
    atom minJuros = 0.0
    atom medJuros = maxJuros / 2.0
    atom minDiferenca = power(0.1, precisao)

    for indice = 1 to maxIteracoes do
        medJuros = (minJuros + maxJuros) / 2.0
        if (maxJuros - minJuros) < minDiferenca then
            return medJuros
        end if
        if jurosParaAcrescimo(medJuros) < acrescimo then
            minJuros = medJuros
        else
            maxJuros = medJuros
        end if
    end for

    return medJuros
end function

procedure main()
    -- calcula e guarda os resultados das funções
    atom pesoTotal = getPesoTotal()
    atom acrescimoCalculado = jurosParaAcrescimo(3.0)
    atom jurosCalculado = acrescimoParaJuros(acrescimoCalculado, 15, 100, 50.0)

    -- imprime os resultados
    puts(STDOUT, "Peso total = ")
    printf(STDOUT, "%2.14f", pesoTotal)
    puts(STDOUT, "\n")
    puts(STDOUT, "Acréscimo = ")
    printf(STDOUT, "%2.14f", acrescimoCalculado)
    puts(STDOUT, "\n")
    puts(STDOUT, "Juros = ")
    printf(STDOUT, "%2.14f", jurosCalculado)
    puts(STDOUT, "\n")
end procedure

main()
